ARCs/arc-0031.md

```diff
@@ -0,0 +1,269 @@
+    deactivate W
+    activate B
+    B->>B: Verify Sig(msg) with PKa and msg
+    Note right of B: if Sig(msg) is valid<br/>then authenticate user
+    B-->>A: Authentication OK/KO
+    deactivate B
+```
+
+The diagram above summarizes the proposed mechanism. We consider the User, **Alice**, owner of an Algorand account <*PKa, SKa*> of which the secret key *SKa* is stored into a **Wallet**.
+
+> A wallet is any type of Algorand wallet, such as hot wallets like <a href="https://www.purestake.com/technology/algosigner/">AlgoSigner</a>, <a href="https://wallet.myalgo.com/">MyAlgo Wallet</a> for browser and mobile wallets used through <a href="https://developer.algorand.org/docs/get-details/walletconnect/">WalletConnect</a>, and cold wallets like the <a href="https://www.ledger.com">Ledger Nano</a>.
+
+Alice authenticates herself to the Verifier **Bob** sending back the digital signature `Sig(msg)` of message `msg` provided by Bob. The mechanism proceeds as follows:
+
+1. Alice initiates an authentication with *PKa* to Bob;
+2. Bob generates a message `msg` to be signed by Alice's *SKa*;
+3. Alice signs `msg` using her wallet; the Wallet inspects and displays the `msg` origin to be sure it came from Bob;
+4. Alice sends back the tuple `<PKa, Sig(msg)>` to Bob;
+5. Bob verifies `Sig(msg)` with Alice's *PKa* and `msg`;
+6. If the signature is valid, Bob authenticates Alice.
+
+The ARC-31 defines a standardized message for authentication, called *Authentication Message*.
+
+### Authentication Message
+
+An *Authentication Message* `auth-msg` is a sequence of bytes representing a message to be signed. The Verifier requests Users to sign an `auth-msg` with their secret keys. Such a message **MUST** include the following information:
+
+- `domain name`: name of the Verifier domain;
+- `Algorand address`: User's *PKa* to be authenticated;
+- `nonce`: a unique/random value generated by the Verifier;
+- `description`: Verifier details or general description;
+- `metadata`: arbitrary message data.
+
+The *Authentication Message* is represented with the following interface:
+
+```typescript
+interface AuthMessage {
+ /** The domain name of the Verifier */
+ domain: string;
+ /** Algorand account to authenticate with*/
+ authAcc: string;
+ /** Unique random nonce generated by the Verifier */
+ nonce: string;
+ /** Optional, description of the Verifier */
+ desc?: string;
+ /** Optional, metadata */
+ meta?: string;
+}
+```
+
+The `auth-msg` **SHOULD** be compliant with [ARC-2](https://github.com/algorandfoundation/ARCs/blob/main/ARCs/arc-0002.md), having the parameter `\<dapp-name\>`=`arc31`, for example:
+
+```json
+arc31:j{
+ "domain": "www.verifierdomain.com",
+ "authAcc": "KTGP47G64KCXWJS64W7SGJNKTHE37TYDCI64USXI3XOYE6ZSH4LCI7NIDA",
+ "nonce": "1234abcde!%",
+ "desc": "The Verifier",
+ "meta": "arbitrary attached data",
+}
+```
+
+The `nonce` field **MUST** be unique for each authentication and **MUST NOT** be used more than once to avoid replay attacks.
```

Comment:
Minimum recommended length?

Reply:
15 characters minimum, which must include upper case and lower case letters and numbers. Nonce checking must be case sensitive. Nonce must be derived from a cryptographically strong PRNG. This would give a search space of 7.82 x 10^26 making it impossible to guess